syndeo-template/src/test/java/com/google/cloud/syndeo/it/PubsubDlqIT.java

```diff
@@ -146,35 +148,30 @@ public void testPubsubToPubsubSuccessful() throws Exception {
     // Publish correct messages to the source queue
     publishCorrectDataToPubsub();
 
-    // Wait for the Syndeo pipeline to move all the data.
-    Thread.sleep(3 * ONE_MINUTE_MILLIS);
-
-    // Wait for a while to check the reults in the pubsub sink
-    {
-      Instant start = Instant.now();
-      while (true) {
-        PullResponse response = pubsubResourceManager.pull(pubsubSinkSubscription, 100);
-        if (response.getReceivedMessagesCount() == MESSAGE_COUNT) {
-          // Test sucessful
-          break;
-        } else if (response.getReceivedMessagesCount() > MESSAGE_COUNT) {
-          throw new AssertionError(
-              String.format(
-                  "Expected at most %s elements to be inserted to Pubsub sink, but found %s.",
-                  MESSAGE_COUNT, response.getReceivedMessagesCount()));
-        } else if (java.time.Duration.between(start, Instant.now())
-                .compareTo(java.time.Duration.ofMinutes(20))
-            > 0) {
-          // If we spent over 20 minutes waiting for the job, then we must exit
-          throw new AssertionError(
-              String.format(
-                  "Expected %s elements to be inserted to Pubsub sink, but found %s after over 20 minutes.",
-                  MESSAGE_COUNT, response.getReceivedMessagesCount()));
-        } else {
-          // Iterate once more
-          Thread.sleep(ONE_MINUTE_MILLIS);
-        }
-      }
+    // Poll the pubsub sink until the messages are written
+    PubsubMessagesCheck pubsubCheck =
+        PubsubMessagesCheck.builder(pubsubResourceManager, pubsubSinkSubscription)
+            .setMinMessages(MESSAGE_COUNT)
+            .setMaxMessages(MESSAGE_COUNT)
+            .build();
+
+    await("Retrieving messages from Pub/Sub sink")
+        .atMost(Duration.ofMinutes(7))
+        .pollInterval(Duration.ofSeconds(15))
+        .until(pubsubCheck::get);
+
+    int msgCount = pubsubCheck.getReceivedMessageList().size();
+
+    if (msgCount > MESSAGE_COUNT) {
+      throw new AssertionError(
+          String.format(
+              "Expected at most %s elements to be inserted to Pubsub sink, but found %s.",
+              MESSAGE_COUNT, msgCount));
+    } else if (msgCount < MESSAGE_COUNT) {
+      throw new AssertionError(
+          String.format(
+              "Expected at least %s elements to be inserted to Pubsub sink, but found %s.",
+              MESSAGE_COUNT, msgCount));
     }
```

Comment:
nit I think await would trigger a test failure and never arrive here. may be good to check though

Reply:
Actually, I don't think it will even go through the (msgCount > MESSAGE_COUNT) condition because of the way I have setup the PubsubMessagesCheck. I have setMinMessages(MESSAGE_COUNT) and setMaxMessages(MESSAGE_COUNT). So await() will end up erroring out if it doesn't get the exact number of messages. I just added this for sanity, I can remove this if you think that would be cleaner.

Reply:
Got it. Makes sense. Maybe the check itself should be only for the minimal, since you don't want to wait 7 minutes if you already know you have more?